### DISCUSSION TITLE
Searching for tracks from online services

### DISCUSSION THREAD

**Discussion by DanWhiting** [hklages/node-red-contrib-sonos-plus#181] [Q&A]
Hi, is it currently possible to search for and play tracks from online services? 

Searching for tracks in the Sonos app displays matching tracks from linked streaming services so that they can be selected to play. When I try searching the library using the provided node, no matching search results are returned (I have no local library and I haven't saved any tracks to mysonos).

I know it's possible to save tracks to mysonos but if possible I'd rather directly search the streaming services.

Thanks!

**Comment by hklages** [comment]
Hi, currently It is not supported. But that's an interesting extension. 
What service / services are most important to you? What kind of search? tracks?
What shall happen, if there are multiple matches?

When I am back from vacation ( December, 10th) I could start work on it.

**Reply by stevegroom** [reply]
I think this answers my question - why do my searches always return nothing. I only have Music & Content services set up - Apple Music, Audible, Sonos Radio and TuneIn. I didn't think a local media library is needed. 

Please add this feature.

**Comment by hklages** [comment]
Hi Steve. 
Search of external services is not yet implemented. I started the first tests in January - but ... 
- each service requests a special login procedure and SONOS does not provide any documentation
- it may change at any time - so maintaining is difficult. 

What is your use case? Assuming there is a "household.search.audible" with payload "searchstring": How would your flow look like?

My approach is to use the SONOS-App, search for interesting songs and add them to My Sonos or add the songs/albums to the SONOS-Queue. 
My Sonos can be searched and saved queues can be played with this pacakge.

**Comment by stevegroom** [comment]
I have just come back to trying to use nodered to create an entertainment dashboard. I’ll hopefully end up with an old ipad showing the nodeRed UI as a remote control with a number of presets. This will make using the system much easier for my elderly mother.

For Audible

Dashboard Button “Resume current audiobook”
V
household.play.audible


Dashboard Button “Play ‘White Nights’”
V
Household.search.audible
V
Add from=latest or from=beginning to the payload.
V
group.play.service

Then for my use case, I converted my Music collection to iTunes Match some years ago. I have an old media library on a NAS, but all the additions to my library are from iTunes Music and only available in DRM protected m4p format. So for all recent additions, I can get them from Apple Music in my Sonos app, but not via node-red-contrib-sonos-plus.

As my home has a mix of smart services, I am using node-red to pull them together into one system. Home Kit, Hue, Alexa, Sonos, Osram, Nest Protect

I see there is a published developer api on the Sonos site – are you using that?

Best
Steve

From: H. Klages ***@***.***>
Date: Thursday, 10 February 2022 at 07:20
To: hklages/node-red-contrib-sonos-plus ***@***.***>
Cc: Steve Groom ***@***.***>, Comment ***@***.***>
Subject: Re: [hklages/node-red-contrib-sonos-plus] Searching for tracks from online services (Discussion #181)

Hi Steve.
Search of external services is not yet implemented. I started the first tests in January - but ...

  *   each service requests a special login procedure and SONOS does not provide any documentation
  *   it may change at any time - so maintaining is difficult.

What is your use case? Assuming there is a "household.search.audible" with payload "searchstring": How would your flow look like?

My approach is to use the SONOS-App, search for interesting songs and add them to My Sonos or add the songs/albums to the SONOS-Queue.
My Sonos can be searched and saved queues can be played with this pacakge.

**Comment by hklages** [comment]
Hi Steve  - thanks for you feedback

- You can do a dashboard with presets already now. Just go to the SONOS App, search/select the audible book, album, track whatever and use the "Add to My Sonos". Then as action in node-red use group.play.mysonos with the titel. Even Audible will work. 
See my "demonstrator" based on Home Assistant or plain node-red with dashboard [here](https://github.com/hklages/node-red-contrib-sonos-plus/wiki#demonstrator) 

- For your other use case it is not that easy. You could create some SONOS Playlist, containing your apple music and launch it with sonos-plus.  But that makes only sense if you don't  change it to often.  But the SONOS App is more comfortable. NodeRED and sonos-plus are good when it comes to processing messages - interaction like search/reduce/select is limited.

For many cases, where the search phrase is know (see your preset use case) you can find a work around. Only if you interactively enter a search phase (via the node-red dashboard) and then find the song in any Music Service, there is currently no solution.  

- For Home Automation you can give Home Assistant a try. Hue, Osram, SONOS, Alexa work out of the box - and for complex flows you can still use Node-RED. Home Assistant has now (2022-02 releas) a Media tab supporting some Music-Services.

- sonos-red is not based on the SONOS developer Api - that is cloud based and need a user registration. sonos-red uses [sonos-ts](https://github.com/svrooij/node-sonos-ts)  

In the next 3 weeks I am going to have a second look at the music services: itunes, apple music and youtube. Lets see what i can do. 

greetings, Henning
